src/shared/quota-util.c

```diff
@@ -40,3 +43,45 @@ int quotactl_path(int cmd, const char *path, int id, void *addr) {
 
         return quotactl_devnum(cmd, devno, id, addr);
 }
+
+int quota_proj_id_is_supported(int fd, uint32_t proj_id, struct dqblk *req) {
+        int r;
+
+        assert(fd >= 0);
+        assert(req);
+
+        r = RET_NERRNO(quotactl_fd(fd, QCMD_FIXED(Q_GETQUOTA, PRJQUOTA), proj_id, req));
+        if (r == -ESRCH || ERRNO_IS_NEG_NOT_SUPPORTED(r) || ERRNO_IS_NEG_PRIVILEGE(r))
+                return false;
+        if (r < 0)
+                return r;
+
+        return true;
+}
+
+int quota_proj_id_set_recursive_verify_exclusive(int fd, uint32_t proj_id) {
+        int r;
+
+        assert(fd >= 0);
+
+        /* Set to top level first because of the case where directories already exist with multiple subdirectories,
+           in which case, number of inodes will be > 1 if applied recursively only */
+        r = set_proj_id(fd, proj_id);
+        if (r < 0)
+                return r;
+
+        /* Confirm only the current inode has the project id (in case of race conditions) */
+        struct dqblk req;
+        r = RET_NERRNO(quotactl_fd(fd, QCMD_FIXED(Q_GETQUOTA, PRJQUOTA), proj_id, &req));
+        if (r < 0)
+                return r;
+
+        if (req.dqb_curinodes != 1)
+                return false;
```

Comment:
i think we should return -EBUSY here, after all if the proj id is already used that's kinda an error